File: app/services/spotify.py
"""Spotify Web API integration for Albricias.

Implements the OAuth 2.0 Authorization Code Flow and fetches monthly
listening data: top tracks, top artists, and recently played tracks.

Required Spotify app scopes:
    user-top-read  user-read-recently-played

Environment variables:
    SPOTIFY_CLIENT_ID
    SPOTIFY_CLIENT_SECRET
    SPOTIFY_REDIRECT_URI   (e.g. http://localhost:5000/admin/spotify/callback)
"""

import os
import base64
import datetime
import urllib.parse
from typing import Any

try:
    import httpx
except ImportError:
    raise ImportError("httpx is required: uv add httpx")

SPOTIFY_AUTH_URL = "https://accounts.spotify.com/authorize"
SPOTIFY_TOKEN_URL = "https://accounts.spotify.com/api/token"
SPOTIFY_API_URL = "https://api.spotify.com/v1"

SCOPES = "user-top-read user-read-recently-played"


# ---------------------------------------------------------------------------
# OAuth helpers
# ---------------------------------------------------------------------------


def _client_id() -> str:
    return os.environ["SPOTIFY_CLIENT_ID"]


def _client_secret() -> str:
    return os.environ["SPOTIFY_CLIENT_SECRET"]


def _redirect_uri() -> str:
    return os.getenv(
        "SPOTIFY_REDIRECT_URI", "http://localhost:5000/admin/spotify/callback"
    )


def _basic_auth_header() -> str:
    raw = f"{_client_id()}:{_client_secret()}"
    return "Basic " + base64.b64encode(raw.encode()).decode()


def get_auth_url(state: str = "") -> str:
    """Build the Spotify /authorize redirect URL.

    Redirect the admin user's browser to this URL to start the OAuth flow.
    """
    params = {
        "client_id": _client_id(),
        "response_type": "code",
        "redirect_uri": _redirect_uri(),
        "scope": SCOPES,
        "show_dialog": "false",
    }
    if state:
        params["state"] = state
    return f"{SPOTIFY_AUTH_URL}?{urllib.parse.urlencode(params)}"


def exchange_code(code: str) -> dict[str, Any]:
    """Exchange an authorization code for access + refresh tokens.

    Returns the full token response dict from Spotify, including:
        access_token, refresh_token, expires_in, scope, token_type
    """
    resp = httpx.post(
        SPOTIFY_TOKEN_URL,
        headers={
            "Authorization": _basic_auth_header(),
            "Content-Type": "application/x-www-form-urlencoded",
        },
        data={
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": _redirect_uri(),
        },
    )
    resp.raise_for_status()
    return resp.json()


def refresh_access_token(refresh_token: str) -> dict[str, Any]:
    """Obtain a new access token using a stored refresh token.

    Returns the fresh token response dict (same shape as exchange_code but
    may omit refresh_token if Spotify did not rotate it).
    """
    resp = httpx.post(
        SPOTIFY_TOKEN_URL,
        headers={
            "Authorization": _basic_auth_header(),
            "Content-Type": "application/x-www-form-urlencoded",
        },
        data={
            "grant_type": "refresh_token",
            "refresh_token": refresh_token,
        },
    )
    resp.raise_for_status()
    return resp.json()


# ---------------------------------------------------------------------------
# Data fetch
# ---------------------------------------------------------------------------


def _api_headers(access_token: str) -> dict:
    return {"Authorization": f"Bearer {access_token}"}


def _parse_timestamp(ts: str | None) -> datetime.datetime | None:
    if not ts:
        return None
    try:
        return datetime.datetime.fromisoformat(ts.replace("Z", "+00:00"))
    except (ValueError, AttributeError):
        return None


def fetch_monthly_listening(access_token: str) -> list[dict[str, Any]]:
    """Fetch Spotify listening data and return normalised activity dicts.

    Calls three endpoints:
    - /me/top/tracks   (short_term ≈ last 4 weeks, limit 20)
    - /me/top/artists  (short_term ≈ last 4 weeks, limit 10)
    - /me/player/recently-played  (last 50 tracks)

    Returns items with keys: event_type, repo (None), title, url, timestamp, raw
    """
    activities: list[dict] = []
    headers = _api_headers(access_token)

    with httpx.Client(timeout=20.0) as client:
        # ------------------------------------------------------------------
        # Top tracks (last ~4 weeks)
        # ------------------------------------------------------------------
        try:
            resp = client.get(
                f"{SPOTIFY_API_URL}/me/top/tracks",
                headers=headers,
                params={"time_range": "short_term", "limit": 20},
            )
            resp.raise_for_status()
            for i, track in enumerate(resp.json().get("items", []), start=1):
                artist_names = ", ".join(
                    a["name"] for a in track.get("artists", [])
                )
                title = f"#{i} {track['name']} — {artist_names}"
                external_url = (
                    track.get("external_urls", {}).get("spotify") or ""
                )
                activities.append(
                    {
                        "event_type": "spotify_track",
                        "repo": None,
                        "title": title[:300],
                        "url": external_url,
                        "timestamp": None,
                        "raw": track,
                    }
                )
        except Exception as exc:
            print(f"[spotify] Top tracks fetch failed: {exc}")

        # ------------------------------------------------------------------
        # Top artists (last ~4 weeks)
        # ------------------------------------------------------------------
        try:
            resp = client.get(
                f"{SPOTIFY_API_URL}/me/top/artists",
                headers=headers,
                params={"time_range": "short_term", "limit": 10},
            )
            resp.raise_for_status()
            for i, artist in enumerate(resp.json().get("items", []), start=1):
                genres = ", ".join(artist.get("genres", [])[:3])
                title = f"#{i} {artist['name']}"
                if genres:
                    title += f" ({genres})"
                external_url = (
                    artist.get("external_urls", {}).get("spotify") or ""
                )
                activities.append(
                    {
                        "event_type": "spotify_artist",
                        "repo": None,
                        "title": title[:300],
                        "url": external_url,
                        "timestamp": None,
                        "raw": artist,
                    }
                )
        except Exception as exc:
            print(f"[spotify] Top artists fetch failed: {exc}")

        # ------------------------------------------------------------------
        # Recently played (up to 50 tracks for genre/vibe context)
        # ------------------------------------------------------------------
        try:
            resp = client.get(
                f"{SPOTIFY_API_URL}/me/player/recently-played",
                headers=headers,
                params={"limit": 50},
            )
            resp.raise_for_status()
            for item in resp.json().get("items", []):
                track = item.get("track", {})
                played_at = _parse_timestamp(item.get("played_at"))
                artist_names = ", ".join(
                    a["name"] for a in track.get("artists", [])
                )
                title = f"{track.get('name', 'Unknown')} — {artist_names}"
                external_url = (
                    track.get("external_urls", {}).get("spotify") or ""
                )
                activities.append(
                    {
                        "event_type": "spotify_played",
                        "repo": None,
                        "title": title[:300],
                        "url": external_url,
                        "timestamp": played_at,
                        "raw": item,
                    }
                )
        except Exception as exc:
            print(f"[spotify] Recently played fetch failed: {exc}")

    return activities

File: web/src/lib/sources/spotify.ts
/**
 * Spotify Web API integration.
 *
 * Ported from `app/services/spotify.py`: the OAuth 2.0 Authorization Code Flow
 * helpers plus the listening-data fetch, generalised from "this month" to an
 * optional `[periodStart, periodEnd)` range.
 *
 * Like the Python original this module is pure service logic — the
 * `/admin/spotify/{connect,callback,disconnect}` route handlers and the
 * `ServiceToken` persistence around them belong to the admin-routes phase.
 *
 * Required Spotify app scopes: `user-top-read user-read-recently-played`.
 *
 * Environment variables:
 *   SPOTIFY_CLIENT_ID
 *   SPOTIFY_CLIENT_SECRET
 *   SPOTIFY_REDIRECT_URI  (e.g. http://localhost:3000/admin/spotify/callback)
 */

import {
  type ActivityItem,
  type Period,
  inPeriod,
  parseTimestamp,
} from "./types";

const SPOTIFY_AUTH_URL = "https://accounts.spotify.com/authorize";
const SPOTIFY_TOKEN_URL = "https://accounts.spotify.com/api/token";
const SPOTIFY_API_URL = "https://api.spotify.com/v1";

export const SPOTIFY_SCOPES = "user-top-read user-read-recently-played";

/** Same 300-char title budget as the Python original. */
const TITLE_MAX = 300;

// ---------------------------------------------------------------------------
// OAuth helpers
// ---------------------------------------------------------------------------

function requireEnv(name: string): string {
  const value = process.env[name];
  if (!value) throw new Error(`${name} is not set.`);
  return value;
}

function redirectUri(): string {
  return (
    process.env.SPOTIFY_REDIRECT_URI ??
    "http://localhost:3000/admin/spotify/callback"
  );
}

function basicAuthHeader(): string {
  const raw = `${requireEnv("SPOTIFY_CLIENT_ID")}:${requireEnv("SPOTIFY_CLIENT_SECRET")}`;
  return `Basic ${Buffer.from(raw, "utf8").toString("base64")}`;
}

/** The token response Spotify returns from both grant types. */
export interface SpotifyTokenResponse {
  access_token: string;
  token_type: string;
  /** Seconds until `access_token` expires. */
  expires_in: number;
  /** Absent when Spotify chooses not to rotate the refresh token. */
  refresh_token?: string;
  scope?: string;
}

/**
 * Build the Spotify `/authorize` redirect URL.
 *
 * Ported from `spotify.py:get_auth_url`. Send the admin's browser here to start
 * the OAuth flow.
 */
export function getAuthUrl(state = ""): string {
  const params = new URLSearchParams({
    client_id: requireEnv("SPOTIFY_CLIENT_ID"),
    response_type: "code",
    redirect_uri: redirectUri(),
    scope: SPOTIFY_SCOPES,
    show_dialog: "false",
  });
  if (state) params.set("state", state);
  return `${SPOTIFY_AUTH_URL}?${params.toString()}`;
}

async function postToken(body: Record<string, string>): Promise<SpotifyTokenResponse> {
  const response = await fetch(SPOTIFY_TOKEN_URL, {
    method: "POST",
    headers: {
      Authorization: basicAuthHeader(),
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: new URLSearchParams(body).toString(),
  });
  if (!response.ok) {
    throw new Error(
      `Spotify token request failed: ${response.status} ${await response.text()}`,
    );
  }
  return (await response.json()) as SpotifyTokenResponse;
}

/**
 * Exchange an authorization code for access + refresh tokens.
 * Ported from `spotify.py:exchange_code`.
 */
export function exchangeCode(code: string): Promise<SpotifyTokenResponse> {
  return postToken({
    grant_type: "authorization_code",
    code,
    redirect_uri: redirectUri(),
  });
}

/**
 * Obtain a fresh access token from a stored refresh token.
 * Ported from `spotify.py:refresh_access_token`.
 *
 * The response may omit `refresh_token` when Spotify does not rotate it.
 */
export function refreshAccessToken(
  refreshToken: string,
): Promise<SpotifyTokenResponse> {
  return postToken({
    grant_type: "refresh_token",
    refresh_token: refreshToken,
  });
}

// ---------------------------------------------------------------------------
// Data fetch
// ---------------------------------------------------------------------------

interface SpotifyArtistRef {
  name: string;
}

interface SpotifyTrack {
  name?: string;
  artists?: SpotifyArtistRef[];
  external_urls?: { spotify?: string };
}

interface SpotifyArtist {
  name?: string;
  genres?: string[];
  external_urls?: { spotify?: string };
}

async function getJson<T>(
  url: string,
  accessToken: string,
  params: Record<string, string>,
): Promise<T> {
  const query = new URLSearchParams(params).toString();
  const response = await fetch(`${url}?${query}`, {
    headers: { Authorization: `Bearer ${accessToken}` },
  });
  if (!response.ok) {
    throw new Error(
      `Spotify request failed: ${response.status} ${await response.text()}`,
    );
  }
  return (await response.json()) as T;
}

export interface SpotifyFetchOptions {
  /** A currently-valid Spotify access token. */
  accessToken: string;
  /**
   * Optional edition period. Only `spotify_played` items carry a timestamp, so
   * only those are filtered; top tracks/artists are always "short_term"
   * (Spotify's trailing ~4 weeks) and have no timestamp to filter on.
   *
   * Omit it to reproduce the Python behaviour of keeping every item.
   */
  period?: Period;
}

/**
 * Fetch Spotify listening data as normalised activity items.
 *
 * Ported from `spotify.py:fetch_monthly_listening`. Calls the same three
 * endpoints, in the same order, with the same limits:
 *   - `/me/top/tracks`              (short_term, limit 20)  → `spotify_track`
 *   - `/me/top/artists`             (short_term, limit 10)  → `spotify_artist`
 *   - `/me/player/recently-played`  (limit 50)              → `spotify_played`
 *
 * Individual endpoint failures are logged and skipped, never thrown.
 */
export async function fetchSpotifyActivity({
  accessToken,
  period,
}: SpotifyFetchOptions): Promise<ActivityItem[]> {
  const activities: ActivityItem[] = [];

  // ---------------------------------------------------------------------
  // Top tracks (last ~4 weeks)
  // ---------------------------------------------------------------------
  try {
    const data = await getJson<{ items?: SpotifyTrack[] }>(
      `${SPOTIFY_API_URL}/me/top/tracks`,
      accessToken,
      { time_range: "short_term", limit: "20" },
    );
    (data.items ?? []).forEach((track, index) => {
      const artistNames = (track.artists ?? []).map((a) => a.name).join(", ");
      activities.push({
        source: "spotify",
        eventType: "spotify_track",
        repo: null,
        title: `#${index + 1} ${track.name} — ${artistNames}`.slice(0, TITLE_MAX),
        url: track.external_urls?.spotify || "",
        timestamp: null,
        raw: track,
      });
    });
  } catch (error) {
    console.error(`[spotify] Top tracks fetch failed: ${describe(error)}`);
  }

  // ---------------------------------------------------------------------
  // Top artists (last ~4 weeks)
  // ---------------------------------------------------------------------
  try {
    const data = await getJson<{ items?: SpotifyArtist[] }>(
      `${SPOTIFY_API_URL}/me/top/artists`,
      accessToken,
      { time_range: "short_term", limit: "10" },
    );
    (data.items ?? []).forEach((artist, index) => {
      const genres = (artist.genres ?? []).slice(0, 3).join(", ");
      let title = `#${index + 1} ${artist.name}`;
      if (genres) title += ` (${genres})`;
      activities.push({
        source: "spotify",
        eventType: "spotify_artist",
        repo: null,
        title: title.slice(0, TITLE_MAX),
        url: artist.external_urls?.spotify || "",
        timestamp: null,
        raw: artist,
      });
    });
  } catch (error) {
    console.error(`[spotify] Top artists fetch failed: ${describe(error)}`);
  }

  // ---------------------------------------------------------------------
  // Recently played (up to 50 tracks for genre/vibe context)
  // ---------------------------------------------------------------------
  try {
    const data = await getJson<{
      items?: { track?: SpotifyTrack; played_at?: string }[];
    }>(`${SPOTIFY_API_URL}/me/player/recently-played`, accessToken, {
      limit: "50",
    });
    for (const item of data.items ?? []) {
      const playedAt = parseTimestamp(item.played_at);
      if (period && !inPeriod(playedAt, period)) continue;
      const track = item.track ?? {};
      const artistNames = (track.artists ?? []).map((a) => a.name).join(", ");
      activities.push({
        source: "spotify",
        eventType: "spotify_played",
        repo: null,
        title: `${track.name ?? "Unknown"} — ${artistNames}`.slice(0, TITLE_MAX),
        url: track.external_urls?.spotify || "",
        timestamp: playedAt,
        raw: item,
      });
    }
  } catch (error) {
    console.error(`[spotify] Recently played fetch failed: ${describe(error)}`);
  }

  return activities;
}

function describe(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}
